### DISCUSSION TITLE
Fetching the Next /31 Prefix in a Larger Prefix

### DISCUSSION THREAD

**Discussion by paulcarre07** [netbox-community/netbox#17937] [Help Wanted!]
Hello everyone,

I'm looking for a way to directly retrieve the next available /31 prefix within a larger prefix in NetBox. Currently, using the following code only returns the next available prefix, but it defaults to a larger size:

```python
Prefix.objects.filter(tags__slug="inter-co").first().get_first_available_prefix()
``` 
In my case, this returns:

```python
IPNetwork('100.67.173.0/29')
``` 
However, I specifically need the next available /31 prefix, not a /29. Is there a built-in way to specify the desired prefix length without additional workaround code?

Here's what I’m currently considering as a workaround:

```python
prefix_length = "/31"  # Define the netmask somewhere as a string.
prefix = Prefix.objects.filter(tags__slug="inter-co").first().get_first_available_prefix()
prefix_31= Prefix(prefix=f"{prefix.network.format()}{prefix_length}")
``` 
If anyone has a cleaner or more direct approach, I'd greatly appreciate it!

**Comment by candlerb** [comment]
If you look at the source of `get_first_available_prefix` in [netbox/ipam/models/ip.py](https://github.com/netbox-community/netbox/blob/v4.1.6/netbox/ipam/models/ip.py#L48-L55) you'll see:

```
    def get_first_available_prefix(self):
        """
        Return the first available child prefix within the prefix (or None).
        """
        available_prefixes = self.get_available_prefixes()
        if not available_prefixes:
            return None
        return available_prefixes.iter_cidrs()[0]
```

Therefore you could call get_available_prefixes(), iterate over them ignoring any which are not /31, and take the first which is /31.

However, I think you might have misunderstood what `available_prefixes` are. Look at the source again:

```
    def get_available_prefixes(self):
        """
        Return all available prefixes within this Aggregate or Prefix as an IPSet.
        """
        params = {
            'prefix__net_contained': str(self.prefix)
        }
        if hasattr(self, 'vrf'):
            params['vrf'] = self.vrf

        child_prefixes = Prefix.objects.filter(**params).values_list('prefix', flat=True)
        return netaddr.IPSet(self.prefix) - netaddr.IPSet(child_prefixes)
```

This returns all the address space which is *not* consumed by a child prefix in the database. So for example, if the prefix is 192.0.2.0/24, and there's also a prefix 192.0.2.0/26 in Netbox, then the "available prefixes" under 192.0.2.0/24 will cover the space 192.0.2.64 .. 192.0.2.255, which is returned as 192.0.2.64/26 and 192.0.2.128/25 (even though those prefixes *don't exist in Netbox*). There are no /31's at all, but of course you could allocate a /31 within either of those.

So it's not really clear to me what you want:
* Do you want to list *child prefixes which already exist in Netbox*, and pick the first which is /31 ?
* Or do you want to enumerate the *unallocated* space, and create a new /31 prefix from that unallocated space?

In the latter case there are various algorithms. A reasonable one (which aims to minimise fragmentation) is to call `get_available_prefixes()`, sort by length (descending), take the prefix with the *longest length* which is <= /31, and if the prefix length is < /31 then take the first /31 out of it.

If there might be multiple prefixes in the database with tag "inter-co" then you should do that across all of them, to minimise fragmentation and to avoid starvation when the first one becomes full.

**Reply by paulcarre07** [reply]
Thank you for the detailed explanation!

To clarify, my goal is to enumerate the unallocated space and create a new /31 prefix from that space. I’ll look into using get_available_prefixes() and following your suggested approach.
